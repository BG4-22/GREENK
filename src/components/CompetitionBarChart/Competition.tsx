import { HStack, VStack } from '@chakra-ui/react';
import React from 'react';
import Chart from './Chart';
import Legend from './Legend';
import YAxisLabel from './YAxixLabel';

interface CompetitionProps {
    data: { name: string; value: number }[];
    unitOfMeasure: string;
}

const Competition: React.VFC<CompetitionProps> = ({ data, unitOfMeasure }) => {
    return (
        <>
            <HStack h="100%" w="100%" justifyContent="flex-start">
                <YAxisLabel text="Andel energi fra miljÃ¸vennlige kilder" />
                <VStack h="100%" w="100%" justifyContent="flex-end">
                    <Chart
                        items={data.map((a) => a.value)}
                        unitOfMeasure={unitOfMeasure}
                    />
                    <Legend items={data.map((a) => a.name)} />
                </VStack>
            </HStack>
        </>
    );
};

export default Competition;
