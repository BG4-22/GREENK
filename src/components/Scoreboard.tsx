import { Button, List, ListItem } from '@chakra-ui/react';
import { useEffect, useState } from 'react';
import { getScores } from '../api';
import { HighscoreEntry } from '../api/types';
import FunfactCard from './funfact-card';

function Scoreboard() {
    const [highscores, setHighscores] = useState<HighscoreEntry[]>([]);
    useEffect(() => {
        getScores().then((data) => setHighscores(data));
    }, []);
    const listItems = highscores.map((item, i) => (
        <ListItem
            w={'65%'}
            margin={'auto'}
            fontSize={'1.1rem'}
            listStylePos={'inside'}
            textAlign={'left'}
            key={`item-${item.name}`}>
            {item.name}
        </ListItem>
    ));

    return (
        <FunfactCard title={'Ukens highscores'}>
            <List
                margin={'0 auto 1rem auto'}
                listStyleType={'decimal'}
                w={'100%'}>
                {listItems}
            </List>
            <Button
                padding={'1rem'}
                margin={'0 auto'}
                bg={'#FFDD85'}
                w={'fit-content'}
                borderRadius={'2rem'}>
                Spill n√•!
            </Button>
        </FunfactCard>
    );

    // return (
    //     <Container>
    //         <Heading size={'md'}>Highscores</Heading>
    //         <Table variant={'striped'}>
    //             <Thead>
    //                 <Tr>
    //                     <Th>Navn</Th>
    //                     <Th>Score</Th>
    //                 </Tr>
    //             </Thead>
    //             <Tbody>
    //                 {highscores.map((highscore) => {
    //                     return (
    //                         <Tr key={highscores.indexOf(highscore) + 1}>
    //                             <Td>{highscore.name}</Td>
    //                             <Td>{highscore.score}</Td>
    //                         </Tr>
    //                     );
    //                 })}
    //             </Tbody>
    //         </Table>
    //     </Container>
    // );
}

export default Scoreboard;
