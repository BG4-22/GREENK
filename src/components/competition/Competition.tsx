import { HStack, Text, VStack } from '@chakra-ui/react';
import React from 'react';
import '../../App.css';
import Chart from './Chart';
import Legend from './Legend';
import YAxisLabel from './YAxisLabel';

interface CompetitionProps {
    data: { name: string; value: number }[];
    unitOfMeasure: string;
}

const Competition: React.VFC<CompetitionProps> = ({ data, unitOfMeasure }) => {
    return (
        <VStack h={'100%'}>
            <Text marginBottom={'1.5em'} fontSize={'1.5vw'} color={'#455b33'}>
                Skolekonkurransen
            </Text>
            <HStack h="80%" w="100%" justifyContent="flex-end">
                <YAxisLabel text="Andel energi fra miljÃ¸vennlige kilder" />
                <VStack h="100%" w="100%" justifyContent="flex-end">
                    <Chart
                        items={data.map((a) => a.value)}
                        unitOfMeasure={unitOfMeasure}
                    />
                    <Legend items={data.map((a) => a.name)} />
                </VStack>
            </HStack>
        </VStack>
    );
};

export default Competition;
