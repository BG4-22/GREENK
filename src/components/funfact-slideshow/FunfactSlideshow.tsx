import { Text } from '@chakra-ui/react';
import React from 'react';
import Carousel from '../carousel/Carousel';
import FunfactCard from '../funfact-card';
export interface FunfactSlideshowPropsI {
    funfacts: string[];
}

const FunfactSlideshow: React.FC<FunfactSlideshowPropsI> = ({ funfacts }) => {
    const elements = funfacts.map((el: string, i: number) => {
        return (
            <Text
                textAlign={'left'}
                width={'90%'}
                margin={'auto'}
                fontSize={'1.5rem'}>
                {el}
            </Text>
        );
    });
    return (
        <FunfactCard title={'Viste du at?'}>
            <Carousel>{elements}</Carousel>
        </FunfactCard>
    );
};

export default FunfactSlideshow;
