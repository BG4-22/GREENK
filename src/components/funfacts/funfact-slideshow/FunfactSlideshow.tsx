import { Box, Text } from '@chakra-ui/react';
import { AnimatePresence, motion } from 'framer-motion';
import { wrap } from 'popmotion';
import React, { FC, useState } from 'react';

import FunfactCard from 'components/funfacts/funfact-card';

const variants = {
    enter: (direction: number) => {
        return {
            x: direction > 0 ? 1000 : -1000,
            opacity: 0,
        };
    },
    center: {
        zIndex: 1,
        x: 0,
        opacity: 1,
    },
    exit: (direction: number) => {
        return {
            zIndex: 0,
            x: direction < 0 ? 1000 : -1000,
            opacity: 0,
        };
    },
};

const swipeConfidenceThreshold = 10000;

const swipePower = (offset: number, velocity: number) => {
    return Math.abs(offset) * velocity;
};

export interface FunfactSlideshowPropsI {
    funfacts: string[];
}

const Carousel: FC<FunfactSlideshowPropsI> = (props) => {
    const { funfacts } = props;
    const [[page, direction], setPage] = useState([0, 0]);
    const funfactIndex = wrap(0, props.funfacts.length, page);

    const paginate = (newDirection: number) => {
        setPage([page + newDirection, newDirection]);
    };
    return (
        <Box
            h={'100%'}
            w={'100%'}
            minH={'100px'}
            position={'relative'}
            flexGrow={1}>
            <Box
                textAlign={'left'}
                width={'90%'}
                margin={'auto'}
                fontSize={'1.5rem'}
                visibility={'hidden'}>
                {funfacts[funfactIndex]}
            </Box>
            <AnimatePresence initial={false} custom={direction}>
                <motion.div
                    style={{
                        position: 'absolute',
                        top: '0px',
                        height: 'fit-content',
                    }}
                    key={page}
                    custom={direction}
                    variants={variants}
                    initial="enter"
                    animate="center"
                    exit="exit"
                    transition={{
                        x: { type: 'spring', stiffness: 300, damping: 30 },
                        opacity: { duration: 0.2 },
                    }}
                    drag="x"
                    dragConstraints={{ left: 0, right: 0 }}
                    dragElastic={1}
                    onDragEnd={(e, { offset, velocity }) => {
                        const swipe = swipePower(offset.x, velocity.x);
                        if (swipe < -swipeConfidenceThreshold) {
                            paginate(1);
                        } else if (swipe > swipeConfidenceThreshold) {
                            paginate(-1);
                        }
                    }}>
                    <Text
                        textAlign={'left'}
                        width={'90%'}
                        margin={'auto'}
                        fontSize={'1.5rem'}>
                        {funfacts[funfactIndex]}
                    </Text>
                </motion.div>
            </AnimatePresence>
        </Box>
    );
};

// inspired by: https://codesandbox.io/s/framer-motion-image-gallery-pqvx3?file=/src/Example.tsx:1598-2616

const FunfactSlideshow: FC<FunfactSlideshowPropsI> = (funfacts) => {
    return (
        <FunfactCard title={'Visste du at?'}>
            <Carousel {...funfacts} />
        </FunfactCard>
    );
};

export default FunfactSlideshow;
