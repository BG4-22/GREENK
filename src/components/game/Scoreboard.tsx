import { Button, List, ListItem } from '@chakra-ui/react';
import { useEffect, useState } from 'react';

import { getHighscores } from '../../services/game';
import { HighscoreEntry } from '../../types/game';
import FunfactCard from '../funfacts/funfact-card';
import GameButton from '../GameButton';

function Scoreboard() {
    const [highscores, setHighscores] = useState<HighscoreEntry[]>([]);
    async function updateHighscores() {
        setHighscores(await getHighscores());
    }

    useEffect(() => {
        updateHighscores();
    }, []);
    const listItems = highscores.map((item, i) => (
        <ListItem
            w={'65%'}
            margin={'auto'}
            fontSize={'1.1rem'}
            listStylePos={'inside'}
            textAlign={'left'}
            key={`$highscore-${i + 1}`}>
            {`${i + 1}. ${item.name}: ${item.score}`}
        </ListItem>
    ));

    return (
        <FunfactCard title={'Ukens highscores'}>
            <List margin={'0 auto 1rem auto'} w={'100%'}>
                {listItems}
            </List>
            <GameButton/>
        </FunfactCard>
    );
}

export default Scoreboard;
