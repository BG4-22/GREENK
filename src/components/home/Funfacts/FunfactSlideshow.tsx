import { Text } from '@chakra-ui/react';
import React from 'react';
import { hashCode } from '../../../utils/hashcode';
import Carousel from '../../common/Carousel';
import Card from '../../common/Card';

/**
 * Slideshow of funfacts inside the card component. Props is sent in the Home component and shown here.
 */

export interface FunfactSlideshowPropsI {
    funfacts: string[];
}

const FunfactSlideshow: React.FC<FunfactSlideshowPropsI> = ({ funfacts }) => {
    const elements = funfacts.map((el: string, i: number) => {
        return (
            <Text key={`${i}-${hashCode(el)}`} sx={textStyle}>
                {el}
            </Text>
        );
    });
    return (
        <Card title={'Visste du at?'}>
            <Carousel withAutomaticSliding navButtons={'small-buttons'}>
                {elements}
            </Carousel>
        </Card>
    );
};

export default FunfactSlideshow;

const textStyle = {
    background: '#fff',
    textAlign: 'center',
    width: '90%',
    fontSize: '1.1rem',
};
