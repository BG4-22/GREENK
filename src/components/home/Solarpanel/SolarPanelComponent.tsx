import { Box } from '@chakra-ui/react';
import React, { useEffect, useState } from 'react';
import getSolarPanelEffect from '../../../api/getSolarPanelEffect';
import Mascot from './Mascot';
import './SolarPanelComponent.css';
import Weather, { WeatherPropsI } from './Weather';

/**
 * This component consists of the Mascot svg-component and the weather component. Based on
 * the effect value, the svg of the sun and cloud will change appearance.
 */

export interface SolarPanelComponentPropsI {
    effect: number;
}

const SolarPanelComponent: React.FC = () => {
    const effect = getSolarPanelEffect();
    // these values should be updated to fit real-time data
    // show the sun if the effect is higher than 6
    const isSunny = (effect: number) => effect > 6;
    // show the cloud if the effect is lower than 12
    const isCloudy = (effect: number) => effect < 12;

    // state to controll the weather
    const [weather, setWeather] = useState<WeatherPropsI>({
        sun: isSunny(effect),
        cloud: isCloudy(effect),
    });

    // setter to update weather
    const updateWeather = (sun: boolean, cloud: boolean): void => {
        setWeather({ sun, cloud });
    };

    useEffect(() => {
        updateWeather(isSunny(effect), isCloudy(effect));
    }, [effect]);

    return (
        <Box
            mb={'20px'}
            h={'100%'}
            w={'100%'}
            pos={'relative'}
            display={'flex'}
            flexDir={'column'}
            justifyContent={'flex-end'}>
            <Weather {...weather} />
            <Mascot effect={effect} />
        </Box>
    );
};

export default SolarPanelComponent;
