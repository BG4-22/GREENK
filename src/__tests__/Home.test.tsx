import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Home from '../pages/Home/Home';
import SolarPanelComponent from '../components/home/Solarpanel/SolarPanelComponent';
import HomeWelcomeComponent from '../pages/Home/Home';
import Scoreboard from '../components/home/Scoreboard';
import Competition from '../components/home/Competition';
import FunfactSlideshow from '../components/home/Funfacts';

describe('<Home />', () => {
    const wrapper = shallow(<Home />);
    it('Should render correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should render every component on the page', () => {
        expect(wrapper.find(<SolarPanelComponent />)).toBeDefined();
        expect(wrapper.find(<HomeWelcomeComponent />)).toBeDefined();
        expect(wrapper.find(<Scoreboard />)).toBeDefined();
        expect(
            wrapper.find(<Competition data={[]} unitOfMeasure="" />)
        ).toBeDefined();
        expect(wrapper.find(<FunfactSlideshow funfacts={[]} />)).toBeDefined();
    });
});
