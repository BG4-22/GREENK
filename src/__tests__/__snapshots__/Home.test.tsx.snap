// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Should render correctly 1`] = `
<Grid
  id="grid"
>
  <ForwardRef
    colSpan={1}
    id="item1"
    rowSpan={1}
  >
    <SolarPanelMascot />
  </ForwardRef>
  <ForwardRef
    colSpan={2}
    id="item2"
    rowSpan={2}
  >
    <HomeWelcomeComponent />
  </ForwardRef>
  <ForwardRef
    colSpan={1}
    id="item3"
    rowSpan={1}
  >
    <Scoreboard />
  </ForwardRef>
  <ForwardRef
    colSpan={1}
    id="item4"
    rowSpan={1}
  >
    <Competition
      data={
        Array [
          Object {
            "name": "Nidarvoll skole",
            "value": 70,
          },
          Object {
            "name": "Gjennomsnitts- skolen i Trondheim",
            "value": 30,
          },
        ]
      }
      unitOfMeasure="%"
    />
  </ForwardRef>
  <ForwardRef
    colSpan={1}
    id="item5"
    rowSpan={1}
  >
    <FunfactSlideshow
      funfacts={
        Array [
          "Hvis man samler all solenergi som treffer jorda på 60 minutter, er det nok energi for hele verden i ett år.",
          "En vindmølle lager nok energi til 1400 hus.  ",
          "Elektrisitet i stømledningene går like fort som lyset.",
          "Fugler kan sitte på en strømledning uten å få støt, men hvis de er borti to samtidig lager de en sluttet krets og kan dø.",
          "Fornybar energi kommer fra kilder som vind, sol, vann og jordvarme. Nidarvoll skole bruker 80% fornybar energi ",
          "Mesteparten av energien vi bruker i verden i dag kommer fra ikke fornybar energi som kull, olje og gass. ",
          "Ikke fornybar energi forurenser masse og verdens kull, olje og gass brukes opp!",
        ]
      }
    />
  </ForwardRef>
</Grid>
`;
